src/Microsoft.TestPlatform.CrossPlatEngine/AttachmentsProcessing/DataCollectorAttachmentsProcessorsFactory.cs

```diff
@@ -0,0 +1,90 @@
+﻿using Microsoft.VisualStudio.TestPlatform.Common.DataCollector;
+using Microsoft.VisualStudio.TestPlatform.Common.ExtensionFramework;
+using Microsoft.VisualStudio.TestPlatform.Common.Logging;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine;
+using Microsoft.VisualStudio.TestPlatform.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.TestRunAttachmentsProcessing
+{
+    internal class DataCollectorAttachmentsProcessorsFactory : IDataCollectorAttachmentsProcessorsFactory
+    {
+        private static Uri CoverageUri = new Uri("datacollector://microsoft/CodeCoverage/2.0");
+        private const string CoverageFriendlyName = "Code Coverage";
+
+        public IReadOnlyDictionary<string, IConfigurableDataCollectorAttachmentProcessor> Create(InvokedDataCollector[] invokedDataCollectors)
+        {
+            IDictionary<string, Tuple<string, IConfigurableDataCollectorAttachmentProcessor>> datacollectorsAttachmentsProcessors = new Dictionary<string, Tuple<string, IConfigurableDataCollectorAttachmentProcessor>>();
+            bool addCodeCoverageAttachmentProcessors = true;
+
+            if (invokedDataCollectors != null)
+            {
+                // We order files by filename descending so in case of the same collector from the same nuget but with different versions, we'll run the newer version.
+                // i.e. C:\Users\xxx\.nuget\packages\coverlet.collector
+                // /3.0.2
+                // /3.0.3
+                // /3.1.0
+                foreach (var invokedDataCollector in invokedDataCollectors.OrderByDescending(d => d.FilePath))
+                {
+                    EqtTrace.Info($"DataCollectorAttachmentsProcessorsFactory: Analyzing data collector attachment processor Uri: {invokedDataCollector.Uri} AssemblyQualifiedName: {invokedDataCollector.AssemblyQualifiedName} FilePath: {invokedDataCollector.FilePath} HasAttachmentProcessor: {invokedDataCollector.HasAttachmentProcessor}");
+
+                    // We'll merge using only one AQN in case of more "same processors" in different assembly.
+                    if (!invokedDataCollector.HasAttachmentProcessor || datacollectorsAttachmentsProcessors.ContainsKey(invokedDataCollector.AssemblyQualifiedName))
+                    {
+                        continue;
+                    }
+
+                    var dataCollectorExtensionManager = DataCollectorExtensionManager.Create(invokedDataCollector.FilePath, TestSessionMessageLogger.Instance);
```

Comment:
>DataCollectorExtensionManager.Create [](http://example.com/codeflow?start=56&length=36)

Do we cache this within the extension manager today?

Reply:
Yes this call will cache the extension.

Reply:
We changed here, we cannot cache this extension because in VS scenario if a user for instance update the runsettings we should load and run the new attachment processor.

Reply:
How is that different from test adapters though? Dont we refresh the cache based on the adapter version?